### PR TITLE
BUG: Fix precise_xstrtod segfault on long exponent

### LINKED ISSUE #38753: REGR: pd.read_csv segfaults with 1.2 (has worked since before pandas 1.0)
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

Wish I could give more information, but this is all I have (not sure I can share the input data):
```
 Fatal Python error: Segmentation fault
Current thread 0x00007f658531a740 (most recent call first):
  File "/usr/local/lib/python3.7/site-packages/pandas/io/parsers.py", line 2056 in read
  File "/usr/local/lib/python3.7/site-packages/pandas/io/parsers.py", line 1052 in read
  File "/usr/local/lib/python3.7/site-packages/pandas/io/parsers.py", line 463 in _read
  File "/usr/local/lib/python3.7/site-packages/pandas/io/parsers.py", line 605 in read_csv
```
Once I added the pin `pandas<1.2` everything works again.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 3e89b4c4b1580aa890023fc550774e63d499da25
python           : 3.7.9.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.4.0-1020-aws
Version          : #29-Ubuntu SMP Wed Jun 14 15:54:52 UTC 2017
machine          : x86_64
processor        : 
byteorder        : little
LC_ALL           : None
LANG             : C.UTF-8
LOCALE           : en_US.UTF-8
pandas           : 1.2.0
numpy            : 1.19.1
pytz             : 2020.5
dateutil         : 2.8.1
pip              : 20.2.2
setuptools       : 49.6.0
Cython           : None
pytest           : 6.2.1
hypothesis       : None
sphinx           : 3.4.1
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.6.2
html5lib         : None
pymysql          : None
psycopg2         : 2.8.6 (dt dec pq3 ext lo64)
jinja2           : 2.11.2
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : 1.5.4
sqlalchemy       : 1.3.22
tables           : None
tabulate         : 0.8.7
xarray           : 0.16.2
xlrd             : None
xlwt             : None
numba            : None
None

</details>

### PR DESCRIPTION
**Pull request by mzeitlin11**
- [x] closes #38753
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

@snowman2 can you check if this branch returns behavior to what you had before 1.2?

### CODE REVIEW

File: pandas/tests/io/parser/test_common.py [hunk 1]
```diff
@@ -1351,6 +1351,30 @@ def test_numeric_range_too_wide(all_parsers, exp_data):
     tm.assert_frame_equal(result, expected)
 
 
+@pytest.mark.parametrize("neg_exp", [-617, -100000, -99999999999999999])
+def test_very_negative_exponent(all_parsers, neg_exp):
+    # GH#38753
+    parser = all_parsers
+    data = f"data\n10E{neg_exp}"
+    for precision in parser.float_precision_choices:
```
Comment (jreback) [inline pandas/tests/io/parser/test_common.py:1359]: you can parameterize on these too

File: pandas/tests/io/parser/test_common.py [hunk 2]
```diff
@@ -1351,6 +1351,30 @@ def test_numeric_range_too_wide(all_parsers, exp_data):
     tm.assert_frame_equal(result, expected)
 
 
+@pytest.mark.parametrize("neg_exp", [-617, -100000, -99999999999999999])
+def test_very_negative_exponent(all_parsers, neg_exp):
+    # GH#38753
```
Comment (jreback) [inline pandas/tests/io/parser/test_common.py:1356]: is this with similar tests?

File: pandas/tests/io/parser/test_common.py [hunk 3]
```diff
@@ -1351,6 +1351,30 @@ def test_numeric_range_too_wide(all_parsers, exp_data):
...
+        expected = DataFrame({"data": [0.0]})
+        tm.assert_frame_equal(result, expected)
+
+
+@pytest.mark.parametrize("exp", [999999999999999999, -999999999999999999])
+def test_too_many_exponent_digits(all_parsers, exp):
+    # GH#38753
+    parser = all_parsers
+    data = f"data\n10E{exp}"
+    for precision in parser.float_precision_choices:
```
Comment (jreback) [inline pandas/tests/io/parser/test_common.py:1370]: same

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
cc @gfyoung

File: pandas/tests/io/parser/test_common.py (hunk 1 above)
Comment (mzeitlin11) [reply pandas/tests/io/parser/test_common.py:1359]: Issue I ran into is that the parsers don't have consistent potential `float_precisions`, eg `PythonParser` only has `None`.  Could parameterize and just assert failure on disallowed combinations of `float_precision` and `parser`, but that seems ugly. Will try to think of a better solution.

File: pandas/tests/io/parser/test_common.py (hunk 2 above)
Comment (mzeitlin11) [reply pandas/tests/io/parser/test_common.py:1356]: Tests above this test overflow, out-of-range numbers that should be parsed as string, edge cases on ends of feasible range, so seemed like the a good spot.

Comment (gfyoung) [reply pandas/tests/io/parser/test_common.py:1356]: This location makes sense for now, but we might be nearing a point where a separate module will do better.

File: pandas/tests/io/parser/test_common.py (hunk 1 above)
Comment (gfyoung) [reply pandas/tests/io/parser/test_common.py:1359]: Some kind of fixture-within-a-fixture functionality would be nice, though not sure if that's feasible...

Comment (mzeitlin11) [reply pandas/tests/io/parser/test_common.py]: Added a new fixture, `all_parsers_all_precisions`. Not the most elegant solution, essentially just precomputes all allowed combos and uses that as the source for the fixture. If you have a better idea please let me know!

**Comment by snowman2** [comment]
> can you check if this branch returns behavior to what you had before 1.2?

`pandas.read_csv` successfully read in the CSV file with this branch without issue.

File: pandas/tests/io/parser/test_common.py [hunk 4]
```diff
@@ -1351,6 +1351,35 @@ def test_numeric_range_too_wide(all_parsers, exp_data):
...
+@pytest.mark.parametrize("exp", [999999999999999999, -999999999999999999])
+def test_too_many_exponent_digits(all_parsers_all_precisions, exp, request):
+    # GH#38753
+    parser, precision = all_parsers_all_precisions
+    data = f"data\n10E{exp}"
+    result = parser.read_csv(StringIO(data), float_precision=precision)
+    if precision == "round_trip":
+        if exp == 999999999999999999 and is_platform_linux():
+            mark = pytest.mark.xfail(reason="On Linux gives object result")
+            request.node.add_marker(mark)
```
Comment (mzeitlin11) [inline pandas/tests/io/parser/test_common.py:1373]: This discrepancy is unrelated, occurs on master, have opened #38794 for it

Comment (jreback) [reply pandas/tests/io/parser/test_common.py:1373]: kk can you put the issue number in the xfail itself.

**Review by jreback — CHANGES_REQUESTED** [review]
small comment, ping on green.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @mzeitlin11 very nice, keep em coming!

**Comment by jreback** [comment]
@meeseeksdev backport 1.2.x
